Adding logs on communication with Controller

jacekchalupka opened the pull request:
Debugging messages are prepended with tags
[Controller-REQUEST] and [Controller-RESPONSE].
Parameter logged:
- for request: method, url, header, body;
- for response: status, header, body.

jacekchalupka committed:
Adding logs on communication with Controller

Debugging messages are prepended with tags
[Controller-REQUEST] and [Controller-RESPONSE].
Parameter logged:
- for request: method, url, header, body;
- for response: status, header, body.

m-kostrzewa commented on logging/log.go:
```diff
@@ -0,0 +1,84 @@
+
+// Function doesn't return error, because it is just for logging.
+// If conversion to json returns error we want to log variable as raw
+func VariableToJSON(variable interface{}) string {
```
Maybe we could make this function private?

jacekchalupka commented:
Logs: https://gist.github.com/jacekchalupka/ae2733f40d3a2a28462ecff88fb1ab52

m-kostrzewa commented on logging/log.go:
```diff
@@ -0,0 +1,84 @@
+			log.Debugln("Cannot read request/response body.", err)
+			buf = []byte("")
+		}
+	}
```
I'm wondering whether we should extract lines 48-69 to a separate function `readBody` or something

m-kostrzewa commented on logging/log.go:
```diff
@@ -0,0 +1,84 @@
+	}
+
+	// NEVER REACHED
+	return ""
```
Instead, we could extract lines 71-80 to a function, e.g. `buildLogMessage`, and just `return buildLogMessage(param, body)` here.

m-kostrzewa commented on logging/log.go:
```diff
@@ -0,0 +1,84 @@
+	case *http.Response:
+		response := param.(*http.Response)
+		response.Body = ioutil.NopCloser(bytes.NewBuffer(buf))
+		return fmt.Sprintf("{ Status: %s, Header: %s, Body: %s }", response.Status, VariableToJSON(response.Header), buf)
```
Maybe it would be better to instead, write to a variable outside of this switch - this would get rid of those nested returns and unreachable code after the switch.

m-kostrzewa commented on client.go:
```diff
@@ -15,6 +15,14 @@ import (
+)
+
+const (
+	loggerRequestPrepend  = "[Controller-REQUEST]"
```
As we talked IRL, maybe it would be better to have 

`logging.HTTPMessage("modulename", req)`

function exposed to users, or something like that.

m-kostrzewa commented on .gitignore:
```diff
@@ -6,6 +6,7 @@
 # Folders
 _obj
 _test
+types
```
This folder is special, as it's generated. I think I would put it at the top of the .gitignore file, while adding a comment why it's so special.

m-kostrzewa commented on logging/log.go:
```diff
@@ -0,0 +1,84 @@
+	return string(jsonOutput)
+}
+
+func HTTPMessageLogger(param interface{}) string {
```
`Logger` in this function name is redundant with name of package.

m-kostrzewa commented on logging/log.go:
```diff
@@ -0,0 +1,84 @@
+		return fmt.Sprintf("[%s]=>[%s] { Header: %s, Body: %s }", request.Method, request.URL.String(), VariableToJSON(request.Header), buf)
+	case *http.Response:
+		response := param.(*http.Response)
+		response.Body = ioutil.NopCloser(bytes.NewBuffer(buf))
```
This transformation from NewBuffer to NopCloser is confusing. If my logger behaved weirdly I would look into this line first looking for possible bug. Could you add a comment here explaining why it's here?

m-kostrzewa commented on logging/log.go:
```diff
@@ -0,0 +1,84 @@
+func HTTPMessageLogger(param interface{}) string {
+
+	if param == nil {
+		return "Empty request/response. Some error occured."
```
Why are we using `log.Debugln("")` in other places for error logging but not here?

Should we `log.Debugln("")` in other places anyways? Such message would show up in debug-level log anyways in the http log message.

jacekchalupka replied:
Because we don't know if it is an error.

m-kostrzewa replied:
Hmm, so maybe let's add this clarification in the code

m-kostrzewa commented on client.go:
```diff
@@ -147,31 +160,42 @@ func (c *Client) httpPut(url string, bodyType string, body io.Reader) (
-	return c.httpClient.Do(req)
+
+	res, err := c.httpClient.Do(req)
+	log.Debugln(loggerResponsePrepend, logging.HTTPMessageLogger(res))
```
Now I'm wondering, whether we should log the request BEFORE calling httpClient.Do, and log the response AFTER we call httpClient.Do. Because http.Do might error out and we will never see the offending request.

To achieve this, we could wrap http.Client.Do into some function, like

```
httpSend(req) err {
  logRequest
  res, err = httpClient.Do(req)
  logResponse // Not sure if we should log it if err happened.
  return err
}
```

m-kostrzewa reviewed (changes requested):
Very good changeset overall. How do you like Go?

Idea for log formatting:

Instead of `[Conroller-REQUEST]` and `[Controller-RESPONSE]`, how about we use the `[POST]=>` and `<=` arrows to signify request and response? Then we would only log the module, like `[Controller]`.

BTW, I would rename it to `[CONTRAIL]` maybe?

jacekchalupka committed:
Debugging messages are prepended with tags
[Controller-REQUEST] and [Controller-RESPONSE].
Parameter logged:
- for request: method, url, header, body;
- for response: status, header, body.

Patch for commit 5d828d78b36ce235631098d88da8106f5dfd3347

jacekchalupka committed:
Fixing nil checking in logging.HTTPmessage

In GO checking if interface is nil fill be false negative.
Should first cast on type or check if param.(type) is nil.

jacekchalupka committed:
Added comment about nil parameter in HTTPMessage

jacekchalupka committed:
Changing logging functions from glog.Errorf to logrus.Debugln

jacekchalupka committed:
Merge branch 'logging' of http://github.com/codilime/contrail-go-api into logging

jacekchalupka commented:
Tested on gerrit: https://review.opencontrail.org/#/c/46435/
